doc/man/hal-disable-polling.1.in
.\" 
.\" hal-disable-polling manual page.
.\" Copyright (C) 2007 David Zeuthen <david@fubar.dk>
.\"
.TH HAL-DISABLE-POLLING 1
.SH NAME
hal-disable-polling \- disable polling on drives with removable media
.SH SYNOPSIS
.PP
.B hal-disable-polling
[options]

.SH DESCRIPTION

\fIhal-disable-polling\fP can be used to to disable and enable media
detection on drives with removable storage. For more information about
both the big picture and specific
.B HAL
properties, refer to the \fIHAL spec\fP which can be found in
.I "@docdir@/spec/hal-spec.html"
depending on the distribution.

.SH OPTIONS
The following options are supported:
.TP
.I "--udi"
The UDI (\fIUnique Device Identifier\fP) of the device object. 
.TP
.I "--device"
The device file of the drive.
.TP
.I "--enable-polling"
Enable polling instead of disabling it.
.TP
.I "--help"
Print out usage.
.TP
.I "--version"
Print the version.

.SH NOTES
.PP
This program requires super user privileges.

.SH RETURN VALUE
.PP
If the requested operation was successful, this program will exit with
exit code 0.

.SH HISTORY
.PP
Polling a storage drive is a necessary evil to detect when the user
inserts or removes media. Human computer interaction studies have
shown that a broad class of users expect their system to react within
a few seconds of this. Thus, the
.I hald
daemon polls through the 
.I hald-addon-storage
addon (one instance for each drive with removable media).

The purpose of the 
.I hald-addon-storage
addon is simply to open the special device file at a regular interval
(either every 2 or every 16 seconds) to check for new media. This
program tries to open the device file using the
.B O_EXCL
option which means that programs like \&\fIcdrecord\fR\|(1) that uses
.B O_EXCL
automatically prevents the 
.I hald-addon-storage
for interferring by continously opening the device file. In addition,
if the drive is locked using HAL (see \&\fIhal-lock\fR\|(1)) the addon
also stops polling.

Unfortunately, polling a storage drive can have adverse side effects
if the hardware and/or device driver for the hardware is
malfunctioning. Additionally, the operating system kernel itself may
offer multiple interfaces for the same device (e.g. \&\fI/dev/sg0\fR\|
and \&\fI/dev/scd0\fR\|) so even
.B O_EXCL
won't work. Also, polling a drive may decrease throughput in certain
(odd and/or broken) configurations; for example, if two IDE drives
shares the same host (master/slave), bus traffic and contention caused
by polling e.g. the optical drive (slave) can reduce throughput to the
hard disk (master) and/or interfere with CD burning on another optical
drive (master). Finally, polling a drive incurs an overhead both in
the host system (processes get woken up often, preventing the CPU to
stay in a deep power saving states) and it may prevent the actual
drive from reaching deep power states as well. As a result, more power
is consumed and this affects battery life for laptops.

Despite the existence of support for asynchronous media change
notification in recent MMC (Multi-Media Commands) specifications,
virtually no optical drives are compliant with the
specification. Fortunately newer SATA ATAPI hardware seems to support
Asynchronous Notification (AN) and at this time of writing (March
2007) work is underway to make both the
.I Linux 
operating system kernel and
.I HAL
take advantage of this.

It is the position of the
.I HAL
team that polling should be avoided at all costs as long as it doesn't
heavily impact the user experience in a negative way. This tool is
provided as a stop-gap measure to use if a system is rendered useless
due to bugs in drivers and/or hardware that is provoked by HAL polling
the drive. If such a bug is encountered it should be reported (see the
.B BUGS
section below) so it can be fixed - historically
.B hald
have triggered a number of bugs in 
.I Linux
storage drivers and related subsystems (such as USB) that have later
been fixed.

.SH BUGS
.PP
Please send bug reports to either the distribution or the HAL
mailing list, see 
.I "http://lists.freedesktop.org/mailman/listinfo/hal"
on how to subscribe.

.SH SEE ALSO
.PP
\&\fIhald\fR\|(8),
\&\fIlshal\fR\|(1),
\&\fIhal-lock\fR\|(1),
\&\fIopen\fR\|(2),
\&\fIhttp://www.t10.org/scsi-3.htm\fR\|,
\&\fIhttps://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=halpolling\fR\|

.SH AUTHOR
Written by David Zeuthen <david@fubar.dk> with a lot of help from many
others.



NEWS
==========
HAL 0.5.10 ""
==========

Requirements for HAL 0.5.10 "":

 - Linux kernel     >= 2.6.17
 - util-linux       >= 2.12r1    (--enable-umount-helper requires patch from RH #188193)
 - bash             >= 2.0
 - udev             >= 089
 - dbus             >= 0.61      (with glib bindings)
 - glib             >= 2.6.0
 - expat            >= 1.95.8
 - hal-info         >= 20070402
 - libusb           >= 0 0.1.10a (optional)
 - pciutils         >= 2.2.3     (optional
 - dmidecode        >= 2.7       (optional)
 - parted           == 1.7.1, 1.8.0, 1.8.1, 1.8.2, 1.86 (optional)
 - cryptsetup-luks  >= 1.0.1     (optional, needs LUKS patches)
 - libsmbios	    >= 0.13.4    (optional for DELL machines, Linux only)
 - dellWirelessCtl  >= 0.13.4    (optional, for Dell machines, must live in /usr/bin/, Linux only)
 - ConsoleKit       >= 0.2.0
 - pm-utils         >= 0.99.2 or newer (optional)


==========
HAL 0.5.9 "Precipice"
==========

Released April 2, 2007.

Requirements for HAL 0.5.9 "Precipice":

 - Linux kernel     >= 2.6.17
 - util-linux       >= 2.12r1    (--enable-umount-helper requires patch from RH #188193)
 - bash             >= 2.0
 - udev             >= 089
 - dbus             >= 0.61      (with glib bindings)
 - glib             >= 2.6.0
 - expat            >= 1.95.8
 - hal-info         >= 20070402
 - libusb           >= 0 0.1.10a (optional)
 - pciutils         >= 2.2.3     (optional
 - dmidecode        >= 2.7       (optional)
 - parted           == 1.7.1, 1.8.0, 1.8.1, 1.8.2, 1.86 (optional)
 - cryptsetup-luks  >= 1.0.1     (optional, needs LUKS patches)
 - libsmbios	    >= 0.13.4    (optional for DELL machines, Linux only)
 - dellWirelessCtl  >= 0.13.4    (optional, for Dell machines, must live in /usr/bin/, Linux only)
 - ConsoleKit       >= 0.2.0
 - pm-utils         >= 0.99.2 or newer (optional)

Contributors to HAL 0.5.9 "Precipice":

Alex Kanavin (1):
      provide textual interface description for USB devices

Andreas Hanke (1):
      fixes for AS_AC_EXPAND usage breaks hal-device-manager

Artem Kachitchkine (5):
      hald/linux/ids -> hald/ids and minor fixes
      fix minor memory leaks
      do not let tools coredump if HAL is not running
      Merge branch 'master' of ssh://git.freedesktop.org/git/hal
      solaris backend

Bastien Nocera (1):
      killswitch infrastructure for Bluetooth and use /dev/sonypi instead of spicctrl

Dan Nicholson (1):
      POSIX shells do not support '==' tests

Daniel Nylander (1):
      added Swedish translation file

Danny Kukawka (63):
      replaced LIBHAL_FREE_DBUS_ERROR with dbus_error_free()
      fixed acpi problems with incorrect low current voltage values
      fixed usage of dbus_error_is_set(error == NULL) due to D-Bus changes
      fixed detection of wireless devices with dscape kernel stack
      fix copy_property for merge in FDI files
      replace usage of g_assert() with error handling
      Merge branch 'master' of ssh://dkukawka@git.freedesktop.org/git/hal
      add hardware dependent/specific alsa sound devices to HAL
      fixed compiler warning
      fixed automake for linux
      fixed typo in 20-storage-methods.fdi
      remove unneeded close() calls
      fixed code docu in libhal
      removed not needed '.in' version of a luks script
      fixed segmentation fault in xen code
      Fix to guarantee that udi get generated before process FDI
      changed error handling for hal-storage-shared.c
      fixed detection of libsmbios to check for >= v0.11.6
      changed D-Bus error if /sbin/cryptsetup is not available
      remove sanitizing passwords for luks and prevent word splitting/path expansion
      fix detection of max. available omnibook brightness levels
      fix get/set brightness for sonypi/spiccrtl
      added new section for deprecated properties to spec
      fixed compiler warnings
      fix .hal-mtab mountpoint handling
      detect path to dmidecode on runtime (fd.o bug #9318)
      added hald-addon-dell-backlight to .gitignore
      fix gtk-doc code documentation
      fix (more) gtk-doc code documentation
      fixed typo in spec
      add copy_property FDI directive for stringlist properties
      fixed compiler warning in addon-dell-backlight
      fixed compiler warning
      fix compiler warning in libhal
      fixed compiler warning in create_cache.c
      fixed commit 882085a0, forgot cast to unsigned long
      typo fix
      fixed uid datatype in CallerInfo to unsigned long
      added new fdi match attribute 'contains_not' for strings and strlist
      added hald-probe-ieee1394-unit to .gitignore
      libhal: validate parameter to be != NULL
      fix expat detection in configure
      fixed code documentation
      added some files to .gitignore
      removed not needed string from translation files
      added support for ALSA MIDI devices
      validate error D-Bus names returned on stderror
      change usb*.version_bcd/usb*.speed_bcd to new double keys
      added hal-system-power-pm-is-supported to .gitignore
      fixed libsmbios search output to correct version
      updated requirements in NEWS file
      void hal-device-manager to open multiple about dialogs
      fix for new pciutils which need to link against zlib
      Merge branch 'master' of ssh://dkukawka@git.freedesktop.org/git/hal
      fix configure for DELL stuff
      fixed quoting and added allowed libparted versions to error output
      added .gitignore to new doc/man dir
      update several .gitignore files
      added optional support for IBM ACPI (hotkey) events.
      added configure option to enable build Toshiba ACPI button addon
      add more cdrom information
      update .gitignore
      updated German translation file

David Woodhouse (1):
      fix crash on PPC Pegasos that don't have PMU, ACPI or APM

David Zeuthen (196):
      post-release version bump to 0.5.9
      fix race condition when unmounting a volume takes a very long time

tools/hal-disable-polling.c
/***************************************************************************
 * CVSID: $Id$
 *
 * hal-disable-polling.c : Disable polling on a drive
 *
 * Copyright (C) 2007 David Zeuthen, <david@fubar.dk>
 *
 * Licensed under the Academic Free License version 2.1
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 **************************************************************************/


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <glib.h>
#include <libhal.h>

/** 
 *  usage:
 *  @argc:                Number of arguments given to program
 *  @argv:                Arguments given to program
 *
 *  Print out program usage. 
 */
static void
usage (int argc, char *argv[])
{
	fprintf (stderr,
                 "\n"
                 "usage : hal-disable-polling [--udi <udi> | --device <device-file>]\n"
                 "                            [--enable-polling]\n"
                 "                            [--help] [--version]\n");
	fprintf (stderr,
                 "\n"
                 "        --udi            Unique Device Id\n"
                 "        --device         Device file\n"
                 "        --enable-polling Enable polling instead of disabling it\n"
                 "        --version        Show version and exit\n"
                 "        --help           Show this information and exit\n"
                 "\n"
                 "This program is provided to make HAL stop polling a drive. Please read.\n"
                 "the entire manual page before using this program.\n"
                 "\n");
}

/** 
 *  main:
 *  @argc:                Number of arguments given to program
 *  @argv:                Arguments given to program
 *
 *  Returns:              Return code
 *
 *  Main entry point 
 */
int
main (int argc, char *argv[])
{
	char *udi = NULL;
	char *device = NULL;
        dbus_bool_t is_version = FALSE;
        dbus_bool_t enable_polling = FALSE;
	DBusError error;
        LibHalContext *hal_ctx;
        FILE *f;
        char *filename;

	if (argc <= 1) {
		usage (argc, argv);
		return 1;
	}

	while (1) {
		int c;
		int option_index = 0;
		const char *opt;
		static struct option long_options[] = {
			{"udi", 1, NULL, 0},
			{"device", 1, NULL, 0},
                        {"enable-polling", 0, NULL, 0},
			{"version", 0, NULL, 0},
			{"help", 0, NULL, 0},
			{NULL, 0, NULL, 0}
		};

		c = getopt_long (argc, argv, "",
				 long_options, &option_index);
		if (c == -1)
			break;

		switch (c) {
		case 0:
			opt = long_options[option_index].name;

			if (strcmp (opt, "help") == 0) {
				usage (argc, argv);
				return 0;
			} else if (strcmp (opt, "version") == 0) {
				is_version = TRUE;
			} else if (strcmp (opt, "udi") == 0) {
				udi = strdup (optarg);
			} else if (strcmp (opt, "device") == 0) {
				device = strdup (optarg);
			} else if (strcmp (opt, "enable-polling") == 0) {
				enable_polling = TRUE;
			}
			break;

		default:
			usage (argc, argv);
			return 1;
			break;
		}
	}

	if (is_version) {
		printf ("hal-disable-polling " PACKAGE_VERSION "\n");
		return 0;
	}

	if (udi == NULL && device == NULL) {
		usage (argc, argv);
		return 1;
	}

	if (udi != NULL && device != NULL) {
		usage (argc, argv);
		return 1;
	}

	dbus_error_init (&error);	
	if ((hal_ctx = libhal_ctx_new ()) == NULL) {
		fprintf (stderr, "error: libhal_ctx_new\n");
		return 1;
	}
	if (!libhal_ctx_set_dbus_connection (hal_ctx, dbus_bus_get (DBUS_BUS_SYSTEM, &error))) {
		fprintf (stderr, "error: libhal_ctx_set_dbus_connection: %s: %s\n", error.name, error.message);
		LIBHAL_FREE_DBUS_ERROR (&error);
		return 1;
	}
	if (!libhal_ctx_init (hal_ctx, &error)) {
		if (dbus_error_is_set(&error)) {
			fprintf (stderr, "error: libhal_ctx_init: %s: %s\n", error.name, error.message);
			dbus_error_free (&error);
		}
		fprintf (stderr, "Could not initialise connection to hald.\n"
				 "Normally this means the HAL daemon (hald) is not running or not ready.\n");
		return 1;
	}

        if (getuid () != 0) {
		fprintf (stderr, "This program requires super user (root) privileges.\n");
                return 1;
        }

        if (device != NULL) {
                char **devices;
                int num_devices;
                int n;

                devices = libhal_manager_find_device_string_match (hal_ctx, "block.device", device, &num_devices, NULL);
                if (devices == NULL) {
                        fprintf (stderr, "Cannot find device %s.\n", device);
                        return 1;
                }

                for (n = 0; devices[n] != NULL; n++) {
                        if (libhal_device_query_capability (hal_ctx, devices[n], "storage", NULL)) {
                                udi = devices[n];
                                break;
                        }
                }

                if (udi == NULL) {
                        fprintf (stderr, "Cannot find storage device %s.\n", device);
                        return 1;
                }

                /* mmmkay, we don't care about leaking the variable devices... mmkay? mmkay! */
        } else {
                if (!libhal_device_exists (hal_ctx, udi, &error)) {
                        fprintf (stderr, "Cannot find device with udi %s.\n", udi);
                        return 1;
                }
                if (!libhal_device_query_capability (hal_ctx, udi, "storage", NULL)) {
                        fprintf (stderr, "Device with udi %s is not a storage device.\n", udi);
                        return 1;
                }
                device = libhal_device_get_property_string (hal_ctx, udi, "block.device", NULL);
                if (device == NULL) {
                        fprintf (stderr, "Device with udi %s does not have block.device set.\n", udi);
                        return 1;
                }
        }

        if (!libhal_device_get_property_bool (hal_ctx, udi, "storage.removable", NULL)) {
                fprintf (stderr, "The given drive don't use removable media so it's not polled anyway.\n");
                return 1;
        }

        filename = g_strdup_printf (PACKAGE_SYSCONF_DIR "/hal/fdi/information/media-check-disable-%s.fdi",
                                    g_basename (udi));

        if (enable_polling) {
                if (libhal_device_get_property_bool (hal_ctx, udi, "storage.media_check_enabled", NULL)) {
                        fprintf (stderr, "Polling is already enabled on the given drive.\n");
                        return 1;
                }

                if (!g_file_test (filename, G_FILE_TEST_EXISTS)) {
                        fprintf (stderr, "Cannot find fdi file %s. Perhaps polling wasn't disabled using this tool?\n", filename);
                        return 1;
                }
                if (unlink (filename) != 0) {
                        fprintf (stderr, "Cannot delete fdi file %s.\n", filename);
                        return 1;
                }
        } else {
                if (!libhal_device_get_property_bool (hal_ctx, udi, "storage.media_check_enabled", NULL)) {
                        fprintf (stderr, "Polling is already disabled on the given drive.\n");
                        return 1;
                }

                if (g_file_test (filename, G_FILE_TEST_EXISTS)) {
                        fprintf (stderr, "The fdi file %s already exist. Cowardly refusing to overwrite it.\n", filename);
                        return 1;
                }

                f = fopen (filename, "w");
                if (f == NULL) {
                        fprintf (stderr, "Cannot open %s for writing.\n", filename);
                        return 1;
                }
                fprintf (f, "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
                         "\n"
                         "<deviceinfo version=\"0.2\">\n"
                         "  <device>\n"
                         "    <match key=\"info.udi\" string=\"%s\">\n"
                         "      <merge key=\"storage.media_check_enabled\" type=\"bool\">false</merge>\n"
                         "    </match>\n"
                         "  </device>\n"
                         "</deviceinfo>\n"
                         "\n", udi);
                fclose (f);
        }

        libhal_device_reprobe (hal_ctx, udi, &error);

        if (enable_polling)
                printf ("Polling for drive %s have been enabled. The fdi file deleted was\n"
                        "  %s\n", device, filename);
        else
                printf ("Polling for drive %s have been disabled. The fdi file written was\n"
                        "  %s\n", device, filename);

        return 0;
}
